Combine meta2json.py and rules.py into dromedary.py

Release-Candidate opened the pull request:
This script combines the two scripts meta2json.py and rules.py into a single one, dromedary.py.

Bug: when running it to generate the JSON, it needs unnecessary additional arguments (`-s` and `-o`).

This is the first step to get a script that 

1. reads a file with an Opam switch description and a list of packages
2. generates the Opam switch in `third-party/_opam`
3. installs the configured packages in this switch 
4. uses this switch as the current one to generate the `BUCK` file 

Before adding these steps, I'd like to have answers to a few questions (see Issues)

vsiles commented:
Before I make more comments, would it be possible to avoid copy/pasting the existing scripts into this new one ? It will make maintenance more complex, and we can't see the changes you did to the original ones. Maybe importing them into your new script would be an option ?

vsiles commented:
I see. I can do the review with that in mind then.

Release-Candidate commented:
Of course it depends on how much `meta2json.py` and `rules.py` are going to change in the future. 
Yes, I know, that's an almost impossible question to answer ;). 
To be on the safe side, I'd then rather change `rules.py` and call both of them from `dromedary.py` - at least for the near future.

vsiles commented:
The life cycle of these scripts (especially rules.py) is being modified every 6 months or so. But usually is quite complex ways, to adapt to changes within opam packages' structure. I'm quite inclined to keep rules.py standalone as we've got our own version for internal usage, which is kept "almost" in sync with this one.
I don't want our internal workflow to block improvement in this project, but it's always good to have it in mind.
Also, seeing, the changes in rules.py would help me review this PR ;)

Release-Candidate commented:
Seems like a plan ;)

Another question before I change rules.py: there are functions (like `_process` and `process_json`) that generate and return lists of `None` (I guess they have some meaning inside of Meta). Should I clean them up or should I let them in there to be compatible with your "real" version of it?

vsiles commented:
You got that right: internally, these are calling some internal "builder" class that does more than just outputing text as we do here. Please do keep them.
We probably can improve the situation by calling some parent method, and having "empty" methods in the OSS version while we have something different internally.

Release-Candidate commented:
Great, thanks, will do.

Release-Candidate commented:
I've changed `dromedary.py` to use both scripts now.

Changes in `meta2json.py`:
- add a command line argument to be able to exclude packages from the list of packages (dune.configurator messed up the BUCK file as it isn't a 'real' package)

Changes in `rules.py`:
- change `standard_library` to `ROOT/lib/ocaml`
- add a comment to the `ocamldebug-exe` resources to warn about `ROOT/lib/ocaml` not working there if `ROOT` is a symlink

Dromedary now only takes the path to the `BUCK` file it generates.
Optional arguments:
- the Opam switch to use. If not set, the currently active one is used
- a list of packages to exclude from the packages added to the generated file
- the root of the relative directory to use, a symlink from the Opam switch to this directory is created. Default: `opam`
- the name of the intermediate JSON file to use. If not set, a temporary one is used (and deleted)
- the name of a single package to process

vsiles commented:
Great, thank you !
I'll review the PR by the end of the week.

vsiles commented on dromedary.py:
```diff
@@ -0,0 +1,329 @@
+    exclude_list = args.exclude
+
+    switch_name = args.switch
+    if switch_name is None:
+        switch_name = opam_switch_name()
+    opam_env = opam_switch_env(switch_name)
```
This should be rewired a bit: you're calling `opam_switch_env` twice in a row if the name is missing. Just it call it once ?

Release-Candidate replied:
Oh, yes. I forgot about that when refactoring the `opam_switch_?` functions.

vsiles reviewed (changes requested):
Overall PR looks good, thank you !
A couple of nitpicks and code quality improvement, but most cosmetics.

Thanks !

Release-Candidate commented on README.md:
```diff
@@ -10,15 +10,18 @@ to support buck2/ocaml for OSS.
-opam switch in the env into a .json file (data.json is provided as an example).
-Running `python3 rules.py` will read the previously generated json file and
-create a BUCK file aimed at OSS released.
+Running `python3 dromedary.py -s SWITCH_ROOT_PATH -o BUCK_PATH` will parse the current opam switch with path `SWITCH_ROOT_PATH` for package information and create a BUCK file `BUCK_PATH`. `dromedary.py` combines the scripts `meta2json.py` and `rules.py`.
```
Oh, actually the README doesn't reflect the changes in the current `dromedary.py` at all. I'm going to rewrite the whole dromedary parts.

Release-Candidate commented on README.md:
```diff
@@ -10,15 +10,18 @@ to support buck2/ocaml for OSS.
+opam switch in the env into a .json file `JSON_PATH` (data.json is provided as an example).
+
+Running `python3 rules.py -i JSON_PATH -s SWITCH_ROOT_PATH -o BUCK_PATH` or `python3 dromedary.py -i JSON_PATH -s SWITCH_ROOT_PATH -o BUCK_PATH` will read the previously generated json file `JSON_PATH` and
+create a BUCK file `BUCK_PATH` aimed at OSS released.
```
Yes. But this description is wrong already.

vsiles replied:
True :D Could you fix it while you're editing the file ? Thanks !

Release-Candidate commented on meta2json.py:
```diff
@@ -315,13 +320,15 @@ def main():
     parser = argparse.ArgumentParser()
     parser.add_argument("-p", "--package")
     parser.add_argument("-o", "--output")
+    parser.add_argument("-e", "--exclude", nargs="*")
```
Yes, of course, will do.

Release-Candidate commented on dromedary.py:
```diff
@@ -0,0 +1,329 @@
+
+
+###############################################################################
+def opam_switch_env(switch: Union[str, None]) -> Dict[str, str]:
```
Oh, yes, thanks.

Release-Candidate commented on dromedary.py:
```diff
@@ -0,0 +1,329 @@
+                args=out.args, code=out.returncode, err=out.stderr.decode("utf-8")
+            )
+        )
+        os.remove(INTERMEDIATE_JSON)  # always remove the temporary file.
```
Yes, that's way better.

Release-Candidate commented on README.md:
```diff
@@ -10,15 +10,18 @@ to support buck2/ocaml for OSS.
 *Disclaimer: rules.py is hardcoded to my (vsiles) current setup, please update it
 to fit yours until we make a proper reusable script ;)*
```
I'd say we can change this disclaimer. It at least works for me too ;)

vsiles replied:
I think it makes sense, yes, with the changes you provide.

vsiles commented:
@Release-Candidate everything looks good. I'll need to import the PR into our system before I can merge, but I see some conflicts. Could you rebase to solve them ? I can try to do it from here but I'd rather you push it to the finish line

Release-Candidate commented:
Thanks, I just did the rebase.

vsiles commented:
The PR is on its way to landing. I had to fight a bit the internal system and apply some formatting rules, but we're almost there.
